src/python/pants/engine/objects.py

```diff
@@ -213,3 +215,29 @@ def non_member_error_message(subject):
     return union.define_instance_of(
         cls, non_member_error_message=staticmethod(non_member_error_message)
     )
+
+
+_K = TypeVar("_K")
+_V = TypeVar("_V")
+
+
+@frozen_after_init
+@dataclass(unsafe_hash=True)
+class HashableDict(Generic[_K, _V]):
```

Comment:
See FrozenDict from Pants.util.frozendict. I think it does what you’re after here.